zebrad/src/components/mempool/storage/verified_set.rs

```diff
@@ -0,0 +1,186 @@
+        self.transactions.len()
+    }
+
+    /// Returns `true` if the set of verified transactions contains the transaction with the
+    /// specified `id.
+    pub fn contains(&self, id: &UnminedTxId) -> bool {
+        self.transactions.iter().any(|tx| &tx.id == id)
+    }
+
+    /// Returns `true` if the given `transaction` has any spend conflicts with transactions in the
+    /// mempool.
+    ///
+    /// Two transactions have a spend conflict if they spent the same UTXO or if they reveal the
+    /// same nullifier.
+    pub fn has_spend_conflicts(&self, unmined_tx: &UnminedTx) -> bool {
+        let tx = &unmined_tx.transaction;
+
+        Self::has_conflicts(&self.spent_outpoints, tx.spent_outpoints())
+            || Self::has_conflicts(&self.sprout_nullifiers, tx.sprout_nullifiers().copied())
+            || Self::has_conflicts(&self.sapling_nullifiers, tx.sapling_nullifiers().copied())
+            || Self::has_conflicts(&self.orchard_nullifiers, tx.orchard_nullifiers().copied())
+    }
+
+    /// Clear the set of verified transactions.
+    ///
+    /// Also clears all internal caches.
+    pub fn clear(&mut self) {
+        self.transactions.clear();
+        self.spent_outpoints.clear();
+        self.sprout_nullifiers.clear();
+        self.sapling_nullifiers.clear();
+        self.orchard_nullifiers.clear();
+    }
+
+    /// Insert a transaction into the set.
+    ///
+    /// # Correctness
+    ///
+    /// The transaction is assumed to not have spend conflicts with any other transaction already
+    /// in the set. Use [`Self::has_spend_conflicts`] to ensure this before calling this method.
+    pub fn insert(&mut self, transaction: UnminedTx) {
+        self.cache_outputs_from(&transaction.transaction);
+        self.transactions.push_front(dbg!(transaction));
+    }
+
+    /// Evict one transaction from the set to open space for another transaction.
+    pub fn evict_one(&mut self) -> Option<UnminedTx> {
+        if self.transactions.is_empty() {
+            None
+        } else {
+            // TODO: use random weighted eviction as specified in ZIP-401 (#2780)
+            let last_index = self.transactions.len() - 1;
+
+            Some(self.remove(last_index))
+        }
+    }
+
+    /// Removes all transactions in the set that matches the `predicate`.
+    ///
+    /// Returns the amount of transactions removed.
+    pub fn remove_all_that(&mut self, predicate: impl Fn(&UnminedTx) -> bool) -> usize {
+        // Clippy is unable to detect that there will be a borrow conflict without the `collect`.
+        #[allow(clippy::needless_collect)]
+        let indices_to_remove: Vec<_> = self
+            .transactions
+            .iter()
+            .enumerate()
+            .filter(|(_, tx)| predicate(tx))
+            .map(|(index, _)| index)
+            .collect();
+
+        let removed_count = indices_to_remove.len();
+
+        for index_to_remove in indices_to_remove.into_iter().rev() {
```

Comment:
Subtle, Correctness-Critical Code: Ordering is important here, but it's not immediately clear why.

Please also add a test that removes multiple transactions at the same time, and makes sure the correct transactions were removed. You might need to increase the mempool size to 3 to do this 🤣 

```suggestion
        // Correctness: remove indexes in reverse order,
        // so earlier indexes still correspond to the same transactions
        for index_to_remove in indices_to_remove.into_iter().rev() {
```

Reply:
The test works, but only if the mempool size is at least 4.
(It could theoretically work with 3, but failures are rare at that size.)

I increased the mempool size, and committed a proptest seed that fails if we remove transactions in ascending index order.